lib/ipam/ipam_block_reader_writer.go

```diff
@@ -347,6 +345,76 @@ func (rw blockReaderWriter) releaseBlockAffinity(ctx context.Context, host strin
 	return nil
 }
 
+func (rw blockReaderWriter) queryAffinity(ctx context.Context, k model.Key, revision string) (*model.KVPair, error) {
+	aff, err := rw.client.Get(ctx, k, revision)
+	if err != nil {
+		return nil, err
+	}
+
+	if aff.Value.(*model.BlockAffinity).State != model.StateDeleted {
+		// We're all good - return the afifnity.
+		return aff, nil
+	}
+
+	// If the affinity is marked as deleted, we need to delete it.
+	if err := rw.deleteAffinity(ctx, aff); err != nil {
+		if _, ok := err.(cerrors.ErrorResourceDoesNotExist); ok {
+			// Already deleted - we're good.
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	// Should return a "does not exist" error.
+	return rw.client.Get(ctx, k, revision)
+}
+
+func (rw blockReaderWriter) updateAffinity(ctx context.Context, aff *model.KVPair) (*model.KVPair, error) {
+	return rw.client.Update(ctx, aff)
+}
+
+func (rw blockReaderWriter) deleteAffinity(ctx context.Context, aff *model.KVPair) error {
+	// Mark the affinity as block deleted.
+	aff.Value.(*model.BlockAffinity).State = model.StateDeleted
```

Comment:
We only need to do this for KDD mode. Would be nice to remove the need for the extra write in etcd, potentially.